azure/scope/machine.go

```diff
@@ -160,27 +160,28 @@ func (m *MachineScope) InitMachineCache(ctx context.Context) error {
 // VMSpec returns the VM spec.
 func (m *MachineScope) VMSpec() azure.ResourceSpecGetter {
 	spec := &virtualmachines.VMSpec{
-		Name:                   m.Name(),
-		Location:               m.Location(),
-		ExtendedLocation:       m.ExtendedLocation(),
-		ResourceGroup:          m.NodeResourceGroup(),
-		ClusterName:            m.ClusterName(),
-		Role:                   m.Role(),
-		NICIDs:                 m.NICIDs(),
-		SSHKeyData:             m.AzureMachine.Spec.SSHPublicKey,
-		Size:                   m.AzureMachine.Spec.VMSize,
-		OSDisk:                 m.AzureMachine.Spec.OSDisk,
-		DataDisks:              m.AzureMachine.Spec.DataDisks,
-		AvailabilitySetID:      m.AvailabilitySetID(),
-		Zone:                   m.AvailabilityZone(),
-		Identity:               m.AzureMachine.Spec.Identity,
-		UserAssignedIdentities: m.AzureMachine.Spec.UserAssignedIdentities,
-		SpotVMOptions:          m.AzureMachine.Spec.SpotVMOptions,
-		SecurityProfile:        m.AzureMachine.Spec.SecurityProfile,
-		DiagnosticsProfile:     m.AzureMachine.Spec.Diagnostics,
-		AdditionalTags:         m.AdditionalTags(),
-		AdditionalCapabilities: m.AzureMachine.Spec.AdditionalCapabilities,
-		ProviderID:             m.ProviderID(),
+		Name:                       m.Name(),
+		Location:                   m.Location(),
+		ExtendedLocation:           m.ExtendedLocation(),
+		ResourceGroup:              m.NodeResourceGroup(),
+		ClusterName:                m.ClusterName(),
+		Role:                       m.Role(),
+		NICIDs:                     m.NICIDs(),
+		SSHKeyData:                 m.AzureMachine.Spec.SSHPublicKey,
+		Size:                       m.AzureMachine.Spec.VMSize,
+		OSDisk:                     m.AzureMachine.Spec.OSDisk,
+		DataDisks:                  m.AzureMachine.Spec.DataDisks,
+		AvailabilitySetID:          m.AvailabilitySetID(),
+		Zone:                       m.AvailabilityZone(),
+		Identity:                   m.AzureMachine.Spec.Identity,
+		UserAssignedIdentities:     m.AzureMachine.Spec.UserAssignedIdentities,
+		SpotVMOptions:              m.AzureMachine.Spec.SpotVMOptions,
+		SecurityProfile:            m.AzureMachine.Spec.SecurityProfile,
+		DiagnosticsProfile:         m.AzureMachine.Spec.Diagnostics,
+		AdditionalTags:             m.AdditionalTags(),
+		AdditionalCapabilities:     m.AzureMachine.Spec.AdditionalCapabilities,
+		CapacityReservationGroupID: ptr.Deref(m.AzureMachine.Spec.CapacityReservationGroupID, ""),
```

Comment:
Instead of doing the `ptr.Deref()` in-line here, it's better to create a function for it as seen for the `ProviderID()` here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/machine.go#L474

Reply:
Added a function for getting the value of `CapacityReservationGroupID` from the spec. Also added unit test for the function.